### Excercise 1.1

---

The above gives about a 1.5% chance of randomly selecting the correct cups. Calculate the minimum number of cups of tea required to ensure the probability of randomly selecting the correct cups is less than or equal to 1%.


The following is adapted from: https://stackoverflow.com/a/4941932

In [14]:
#Math Module
import math

In [15]:
#No of ways selecting 4 cups from 8
math.comb(8,4)

70

In [16]:
#No of ways selecting 5 cups from 10
math.comb(10,5)

252

*As you can see from above there are 252 possible outcomes*

In [17]:
#Probability of randomoly selecting correct 5 cups is:
1.0 / math.comb(10,5)

0.003968253968253968

So if the experiment is done with 10 cups, five with milk first and five with tea first, then selecting correctly randomly is less than 1%

In [18]:
#No of ways selecting 6 cups from 12
math.comb(12,6)

924

In [19]:
#Probability of randomoly selecting correct 6 cups is:
1.0 / math.comb(12,6)

0.0010822510822510823

This too is also less than 1%

Bonus: How many would be required if you were to let the taster get one cup wrong while maintaining the 1% threshold?

### Excercise 1.2

---

Use scipy's version of Fisher's exact test to simulate the Lady Tasting Tea problem.

In [20]:
#import the relevant packages
import scipy.stats as ss
import numpy as np
import pandas as pd
import scipy
import scipy.special

Code was taken and adapted from: https://towardsdatascience.com/fishers-exact-test-from-scratch-with-python-2b907f29e593

In [21]:
ar=np.array([[4.0, 0.0],[0.0, 4.0]])    
df=pd.DataFrame(ar, columns=["Guess Milk", "Guess Tea"])
df.index=["Milk", "Tea"] 
df 

Guess Milk  Guess Tea
Milk         4.0        0.0
Tea          0.0        4.0

In [22]:
df2=df.copy()
df2.loc['Column Total']= df2.sum(numeric_only=True, axis=0)
df2.loc[:,'Row Total'] = df2.sum(numeric_only=True, axis=1)
df2

Guess Milk  Guess Tea  Row Total
Milk                 4.0        0.0        4.0
Tea                  0.0        4.0        4.0
Column Total         4.0        4.0        8.0

In [23]:
n=df2.at["Column Total", "Row Total"]  # grand total 

exp=df2.copy()               # create dataframe with expected frequencies
for x in exp.index[0:-1]:
    for y in exp.columns[0:-1]:
        # round expected values to nearest hundredths:
        v= (((df2.at[x, "Row Total"]) * (df2.at["Column Total", y])   )   /n ).round(2) 
        exp.at[x,y]=float(v)

exp

Guess Milk  Guess Tea  Row Total
Milk                 2.0        2.0        4.0
Tea                  2.0        2.0        4.0
Column Total         4.0        4.0        8.0

In [24]:
def p(a): 
    v=(scipy.special.binom(int(df2.iloc[0,2]), a) * scipy.special.binom(int(df2.iloc[1,2]), (int(df2.iloc[2,0])-a)) )/scipy.special.binom(n, int(df2.iloc[2,0]))
    return v

p(1) # if we try "a=1" we get the following probability ... 

0.22857142857142856

In [25]:
p_observed=p(7) # In our contingency table, a was equal to 7.

p_list=[]
for i in range(int(df2.iloc[0,2])  + 1  ): # calculate p(a) for every possible table we can get given the fixed margins...
#... this ranges from "9 choose 0" to "9 choose 9" so we should get 10 possible tables and their respective probabilities..
    if p(i)<=p_observed:
        p_list.append(p(i))     # append these probabilites to p_list only if <= p_observed
        
p_val=np.sum(p_list) # the sum of this list corresponds to the p-value 
p_val 

0.0

In [26]:
oddsratio, pvalue = ss.fisher_exact([[4, 0],[0, 4]])  
pvalue

0.028571428571428536

### Scipy


Scipy is an additional package for statical analysis with scipy.stats being an sub-package. The main purpose of this package is to show probability distributions and statical processes. It contains functions for either continuous or discreet variables with bionamial, uniform and continous distributions. https://data-flair.training/blogs/scipy-statistical-functions/

### Excercise 1.2

---
Take the code from the Examples section of the scipy stats documentation for independent samples t-tests, add it to your own notebook and add explain how it works using MarkDown cells and code comments. Improve it in any way you think it could be improved.

### What is a T-Test?

---

Also known as a Student test is used to for estimating the means of one or two populations using hypothesis testing. It can help evaluate one single group or two groups whether they are independnt or not. https://www.jmp.com/en_ch/statistics-knowledge-portal/t-test.html


The purpose of a T-Test

- Is there a significant difference between two sets of data?